python/ray/util/spark/cluster_init.py

```diff
@@ -0,0 +1,758 @@
+            # The exception might be raised in following cases:
+            #  1. The background job raises unexpected exception (i.e. ray cluster dies
+            #    unexpectedly)
+            #  2. User explicitly orders shutting down the ray cluster.
+            #  3. On Databricks runtime, when a notebook is detached, it triggers python REPL
+            #    `onCancel` event, cancelling the background running spark job
+            #  For case 1 and 3, only ray workers are killed, but driver side ray head might still
+            #  be running and the ray context might be in connected status. In order to disconnect
+            #  and kill the ray head node, a call to `ray_cluster_handler.shutdown()` is performed.
+            if not ray_cluster_handler.spark_job_is_canceled:
+                ray_cluster_handler.shutdown(cancel_background_job=False)
+                ray_cluster_handler.background_job_exception = e
+
+    try:
+        threading.Thread(
+            target=inheritable_thread_target(background_job_thread_fn), args=()
+        ).start()
+
+        # wait background spark task starting.
+        for _ in range(_BACKGROUND_JOB_STARTUP_WAIT):
+            time.sleep(1)
+            if ray_cluster_handler.background_job_exception is not None:
+                raise RuntimeError(
+                    "Ray workers failed to start."
+                ) from ray_cluster_handler.background_job_exception
+
+        start_hook.on_spark_background_job_created(spark_job_group_id)
+        return ray_cluster_handler
+    except Exception:
+        # If driver side setup ray-cluster routine raises exception, it might result in part of ray
+        # processes has been launched (e.g. ray head or some ray workers have been launched),
+        # calling `ray_cluster_handler.shutdown()` to kill them and clean status.
+        ray_cluster_handler.shutdown()
+        raise
+
+
+_active_ray_cluster = None
+
+
+@PublicAPI(stability="alpha")
+def init_ray_cluster(
+    num_worker_nodes: int,
+    object_store_memory_per_node: Optional[int] = None,
+    head_options: Optional[Dict] = None,
+    worker_options: Optional[Dict] = None,
+    ray_temp_root_dir: Optional[str] = None,
+    safe_mode: Optional[bool] = True,
+) -> None:
+    """
+    Initialize a ray cluster on the spark cluster by starting a ray head node in the spark
+    application's driver side node.
+    After creating the head node, a background spark job is created that
+    generates an instance of `RayClusterOnSpark` that contains configuration for the ray cluster
+    that will run on the Spark cluster's worker nodes.
+    After a ray cluster initialized, your python process automatically connect to the ray cluster,
+    you can call `ray.util.spark.shutdown_ray_cluster` to shut down the ray cluster.
+    Note: If the active ray cluster haven't shut down, you cannot create a new ray cluster.
```

Comment:
Yes this is by design.
Because the public API `init_ray_cluster` it will create a cluster and then connect to it,
one python process can only connect to one cluster at a time.
To make API simple, we only support running one cluster at a time